sycl/test-e2e/Matrix/joint_matrix_int8_vnni_impl.hpp

```diff
@@ -78,50 +68,74 @@ void matrix_multiply(big_matrix<T1, NUM_ROWS_C, NUM_COLS_C> &C,
    }).wait();
 }
 
-void int8_row_vnni_reformat(int8_t *_in, int8_t *_out, int K, int N,
-                            int stride_in) {
+template <typename T, size_t N, size_t K, size_t VNNI>
+void row_vnni_reformat(T *_in, T *_out, size_t stride_in) {
   // find the old index, new index, and copy element.
-  //(K, N) => (k/4, N*4)
-  // idx in 2d: (i,j)=>(i/4, j*4+i%4)
+  //(K, N) => (k/VNNI, N*VNNI)
+  // idx in 2d: (i,j)=>(i/VNNI, j*VNNI+i%VNNI)
   // linear idx:
   for (int i = 0; i < K; ++i) {
     for (int j = 0; j < N; ++j) {
       size_t oldindex = i * stride_in + j;
-      size_t newindex = (i / 4) * N * 4 + j * 4 + i % 4;
+      size_t newindex = (i / VNNI) * N * VNNI + j * VNNI + i % VNNI;
       _out[newindex] = _in[oldindex];
     }
   }
 }
 
-int main() {
+template <typename TResult, typename T, size_t VNNI, size_t TM, size_t TN,
+          size_t TK>
+void test() {
   static constexpr size_t MATRIX_M = TM * 2;
   static constexpr size_t MATRIX_N = TN * 2;
   static constexpr size_t MATRIX_K = TK * 2;
-  int8_t A[MATRIX_M][MATRIX_K];
-  int8_t B[MATRIX_K][MATRIX_N];
-  int8_t Bvnni[MATRIX_K / 4][MATRIX_N * 4];
-  int32_t C[MATRIX_M][MATRIX_N];
-  int32_t D[MATRIX_M][MATRIX_N];
+  T A[MATRIX_M][MATRIX_K];
+  T B[MATRIX_K][MATRIX_N];
+  T Bvnni[MATRIX_K / VNNI][MATRIX_N * VNNI];
+  TResult C[MATRIX_M][MATRIX_N];
+  TResult D[MATRIX_M][MATRIX_N];
 
-  matrix_fill(MATRIX_M, MATRIX_K, (int8_t *)A,
-              [](int i, int j) { return i + j; });
-  matrix_fill(MATRIX_K, MATRIX_N, (int8_t *)B,
+  matrix_fill(MATRIX_M, MATRIX_K, (T *)A, [](int i, int j) { return i + j; });
+  matrix_fill(MATRIX_K, MATRIX_N, (T *)B,
               [](int i, int j) { return i + j * 2; });
-  matrix_fill(MATRIX_M, MATRIX_N, (int32_t *)C, 0);
-  matrix_fill(MATRIX_M, MATRIX_N, (int32_t *)D, 0);
+  matrix_fill(MATRIX_M, MATRIX_N, (TResult *)C, 0);
+  matrix_fill(MATRIX_M, MATRIX_N, (TResult *)D, 0);
+
+  big_matrix<TResult, MATRIX_M, MATRIX_N> MC((TResult *)&C);
+  big_matrix<TResult, MATRIX_M, MATRIX_N> MD((TResult *)&D);
+  big_matrix<T, MATRIX_M, MATRIX_K> MA((T *)&A);
+  big_matrix<T, MATRIX_K / VNNI, MATRIX_N * VNNI> MB((T *)&B);
```

Comment:
Hmm, seems there was a bug in the test... if we use row major layout for Matrix B, then it should not be VNNI'ed:
```suggestion
  big_matrix<T, MATRIX_K, MATRIX_N> MB((T *)&B);
```